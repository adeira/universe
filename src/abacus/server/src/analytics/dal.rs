use serde::Deserialize;
use uuid::Uuid;

use crate::arango::{resolve_aql, resolve_aql_vector, ConnectionPool};
use crate::price::Price;

#[derive(Deserialize, Clone, juniper::GraphQLObject, Debug)]
pub(in crate::analytics) struct AnalyticsSoldProductInfo {
    product_id: String,
    product_name: String,
    product_units: i32,
}

#[derive(Deserialize)]
pub struct Redirect {
    _id: String,
    _key: String,
    redirects_to: String,
    description: String,
    hits: i32,
}

#[juniper::graphql_object]
impl Redirect {
    fn id(&self) -> juniper::ID {
        juniper::ID::from(self._id.to_owned())
    }

    /// UUID is the ID used for redirects, for example: https://â€¦/redirect/:uuid
    fn uuid(&self) -> String {
        // Technically, we use the ArangoDB _key as a UUID but we don't want to expose this
        // implementation detail directly.
        self._key.to_owned()
    }

    pub(crate) fn redirects_to(&self) -> String {
        self.redirects_to.to_owned()
    }

    fn description(&self) -> String {
        self.description.to_owned()
    }

    fn hits(&self) -> i32 {
        self.hits.to_owned()
    }
}

/// This function tries to find a link in the database based on the specified identifier and
/// returns it. Additionally, it records a "hit" which is basically of counter of the requests.
pub(in crate::analytics) async fn get_link_and_record_hit(
    pool: &ConnectionPool,
    uuid: &Uuid,
) -> anyhow::Result<Redirect> {
    resolve_aql(
        pool,
        r#"
            LET redirect = DOCUMENT("analytics_redirects", @uuid)
            UPDATE redirect WITH { hits: redirect.hits + 1 } IN analytics_redirects
            RETURN NEW
        "#,
        hashmap_json![
            "uuid" => uuid.hyphenated().encode_lower(&mut Uuid::encode_buffer()),
        ],
    )
    .await
}

pub(in crate::analytics) async fn get_redirect_hits(
    pool: &ConnectionPool,
) -> anyhow::Result<Vec<Redirect>> {
    resolve_aql_vector(
        pool,
        r#"
            FOR redirect IN analytics_redirects
              SORT redirect.hits DESC
              RETURN redirect
        "#,
        hashmap_json![],
    )
    .await
}
