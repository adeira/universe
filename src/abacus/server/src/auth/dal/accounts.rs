use crate::arango::{resolve_aql, resolve_aql_vector, Document};
use crate::auth::account::Account;
use crate::auth::users::AnyUser;

pub(crate) async fn find_user_accounts(
    pool: &crate::arango::ConnectionPool,
    user: &AnyUser,
) -> anyhow::Result<Vec<Document<Account>>> {
    resolve_aql_vector(
        pool,
        r#"
            FOR account IN 1 OUTBOUND @user_id
                GRAPH "users_in_accounts"
                RETURN account
        "#,
        hashmap_json![
            "user_id" => user.id()
        ],
    )
    .await
}

pub(crate) async fn create_new_account(
    pool: &crate::arango::ConnectionPool,
    user: &AnyUser,
) -> anyhow::Result<Document<Account>> {
    resolve_aql(
        pool,
        r#"
            LET new_account = FIRST(
                INSERT {
                  is_active: true,
                  name: "Unnamed Account"
                } INTO accounts
                RETURN NEW
            )

            INSERT {
              _from: @user_id,
              _to: new_account._id,
            } INTO user_accounts

            RETURN new_account
        "#,
        hashmap_json![
            "user_id" => user.id()
        ],
    )
    .await
}
