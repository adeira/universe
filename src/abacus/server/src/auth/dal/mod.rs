pub mod accounts;
pub mod sessions;
pub mod users;
