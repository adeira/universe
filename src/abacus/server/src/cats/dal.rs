use crate::arango::{resolve_aql_vector, ConnectionPool};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct CatInfo {
    _id: String,
    _key: String,
    name: String,
    description: String,
    order: i32,
    can_be_adopted: Option<bool>,
    date_adoption: Option<String>,
    date_deworming: Option<String>,
    date_castration: Option<String>,
    date_vaccination_rabies: Option<String>,
    date_vaccination_triple_felina: Option<String>,
    date_vaccination_cuadruple_felina: Option<String>,
}

// TODO: expose "birth" (approx. age)
#[juniper::graphql_object]
impl CatInfo {
    fn id(&self) -> juniper::ID {
        juniper::ID::from(self._id.to_owned())
    }

    fn name(&self) -> String {
        self.name.to_owned()
    }

    fn description(&self) -> String {
        self.description.to_owned()
    }

    /// Order in which the cat was admitted to KOCHKA Café.
    fn order(&self) -> i32 {
        self.order.to_owned()
    }

    /// Most of the cats can be adopted but some cannot. This flag distinguishes exactly this.
    fn can_be_adopted(&self) -> Option<bool> {
        self.can_be_adopted.to_owned()
    }

    /// When was the cat adopted (when it leaved KOCHKA Café).
    fn date_of_adoption(&self) -> Option<String> {
        self.date_adoption.to_owned()
    }

    /// When was the cat dewormed for the last time.
    fn date_of_deworming(&self) -> Option<String> {
        self.date_deworming.to_owned()
    }

    /// When was the cat castrated.
    fn date_of_castration(&self) -> Option<String> {
        self.date_castration.to_owned()
    }

    fn date_of_vaccination_rabies(&self) -> Option<String> {
        self.date_vaccination_rabies.to_owned()
    }

    /// Felocell 3 and similar.
    fn date_of_vaccination_triple_felina(&self) -> Option<String> {
        self.date_vaccination_triple_felina.to_owned()
    }

    /// Purevax Feline 4 and similar.
    fn date_of_vaccination_cuadruple_felina(&self) -> Option<String> {
        self.date_vaccination_cuadruple_felina.to_owned()
    }
}

#[derive(juniper::GraphQLInputObject)]
pub struct AllCatsFilter {
    /// When `true` returns only adopted cats. When `false` returns only cats available for adoption.
    adopted: bool,
}

pub(in crate::cats) async fn list_all_cats(
    pool: &ConnectionPool,
    filter: &Option<AllCatsFilter>,
) -> anyhow::Result<Vec<CatInfo>> {
    match filter {
        Some(filter) => {
            resolve_aql_vector(
                pool,
                r#"
                  FOR cat IN cats
                    FILTER IS_NULL(cat.date_adoption) != @adopted
                    SORT cat.order ASC
                    RETURN cat
                "#,
                hashmap_json![
                    "adopted" => filter.adopted,
                ],
            )
            .await
        }
        None => {
            resolve_aql_vector(
                pool,
                r#"
                  FOR cat IN cats
                    SORT cat.order ASC
                    RETURN cat
                "#,
                hashmap_json![],
            )
            .await
        }
    }
}
