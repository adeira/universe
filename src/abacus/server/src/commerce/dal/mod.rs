pub(in crate::commerce) mod orders;
pub(in crate::commerce) mod product_addons;
pub(in crate::commerce) mod product_categories;
pub(in crate::commerce) mod products;

#[cfg(test)]
mod product_categories_tests;
#[cfg(test)]
mod products_tests;
