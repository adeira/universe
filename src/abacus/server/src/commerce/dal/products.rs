use crate::arango::{resolve_aql, resolve_aql_vector, ConnectionPool};
use crate::commerce::model::products::{
    PriceSortDirection, Product, ProductMultilingualInput, ProductMultilingualInputVisibility,
};
use crate::images::Image;
use crate::locale::SupportedLocale;
use serde_json::json;

/// Takes care of creating the product inside ArangoDB.
pub(in crate::commerce) async fn create_product(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_multilingual_input: &ProductMultilingualInput,
    images: &[Image],
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@client_locale]

            INSERT {
              images: @product_images,
              unit_label: "product_units/piece", // TODO: dynamic `unit_label`
              is_published: false,
              visibility: @product_visibility,
              addons: @product_addons,
              created: DATE_ISO8601(DATE_NOW()),
              updated: DATE_ISO8601(DATE_NOW()),
              price: {
                unit_amount: @product_price_unit_amount,
                unit_amount_currency: "MXN",
              },
              translations: @translations
            } INTO products
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @client_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_images" => images,
            "product_visibility" => product_multilingual_input.visibility(),
            "product_addons" => product_multilingual_input.addons(),
            "product_price_unit_amount" => product_multilingual_input.price.unit_amount,
            "translations" => product_multilingual_input.translations,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn update_product(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_key: &str,
    product_revision: &str,
    product_multilingual_input: &ProductMultilingualInput,
    images: &[Image],
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@client_locale]

            UPDATE {
              _key: @product_key,
              _rev: @product_rev,
              images: @product_images,
              visibility: @product_visibility,
              addons: @product_addons,
              updated: DATE_ISO8601(DATE_NOW()),
              price: {
                unit_amount: @product_price_unit_amount,
                unit_amount_currency: "MXN",
              },
              translations: @translations
            } IN products OPTIONS { ignoreRevs: false }
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @client_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_key" => product_key,
            "product_rev" => product_revision,
            "product_images" => images,
            "product_visibility" => product_multilingual_input.visibility(),
            "product_addons" => product_multilingual_input.addons(),
            "product_price_unit_amount" => product_multilingual_input.price.unit_amount,
            "translations" => product_multilingual_input.translations,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn publish_product(
    pool: &ConnectionPool,
    product_key: &str,
    client_locale: &SupportedLocale,
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@client_locale]

            UPDATE {
              _key: @product_key,
              is_published: true,
            } IN products
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @client_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_key" => product_key,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn unpublish_product(
    pool: &ConnectionPool,
    product_key: &str,
    client_locale: &SupportedLocale,
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@client_locale]

            UPDATE {
              _key: @product_key,
              is_published: false,
            } IN products
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @client_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_key" => product_key,
        ],
    )
    .await
}

/// Returns a single product (or error). It is possible to search for the products by their
/// ArangoDB `_key` as well as by their `_id`.
pub(in crate::commerce) async fn get_product_by_key_or_id(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_key_or_id: &str,
    product_published_only: &bool,
) -> anyhow::Result<Product> {
    let product_vector = get_products_by_keys_or_ids(
        pool,
        client_locale,
        &[product_key_or_id.to_string()],
        product_published_only,
    )
    .await?;

    match product_vector.first() {
        Some(product) => Ok(product.to_owned()),
        None => anyhow::bail!("database didn't return any product"),
    }
}

pub(in crate::commerce) async fn get_products_by_keys_or_ids(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_keys_or_ids: &[String],
    product_published_only: &bool,
) -> anyhow::Result<Vec<Product>> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql_vector(
        pool,
        r#"
            LET products = DOCUMENT(products, @product_keys_or_ids)
            FOR product IN products
              FILTER @product_published_only ? product.is_published == @product_published_only : true

              LET t = FIRST(
                FOR t IN product.translations
                  FILTER t.name != null AND t.locale == @client_locale
                  RETURN t
              )

              RETURN MERGE(
                product,
                { unit_label: DOCUMENT(product.unit_label)[@client_locale] },
                { name: t.name, description: t.description }
              )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_keys_or_ids" => product_keys_or_ids,
            "product_published_only" => product_published_only,
        ]
    ).await
}

/// Performs search of products based on the specified criteria and returns products with merged
/// translations based on the eshop language.
///
/// TODO(004) - integration tests
pub(in crate::commerce) async fn search_products(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    price_sort_direction: &PriceSortDirection,
    search_all: &bool,
    visibility: &Option<ProductMultilingualInputVisibility>,
) -> anyhow::Result<Vec<Option<Product>>> {
    let sort_direction = match price_sort_direction {
        PriceSortDirection::LowToHigh => "ASC",
        PriceSortDirection::HighToLow => "DESC",
    };

    let visibility = match visibility {
        Some(visibility) => json!(visibility),
        None => json!(null),
    };

    resolve_aql_vector(
        pool,
        r#"
            FOR product IN products
              FILTER @search_all == true ? true : (product.is_published IN [true])
              FILTER @visibility == null ? true : (@visibility IN product.visibility)
              SORT product.price.unit_amount @price_sort_direction

              LET t = FIRST(
                FOR t IN product.translations
                  FILTER t.name != null AND t.locale == @client_locale
                  RETURN t
              )

              RETURN MERGE(
                product,
                { unit_label: DOCUMENT(product.unit_label)[@client_locale] },
                { name: t.name, description: t.description }
              )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "search_all" => search_all,
            "price_sort_direction" => sort_direction,
            "visibility" => visibility,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn search_products_in_categories(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    price_sort_direction: &PriceSortDirection,
    categories: &[juniper::ID],
    search_all: &bool,
    visibility: &Option<ProductMultilingualInputVisibility>,
) -> anyhow::Result<Vec<Option<Product>>> {
    let sort_direction = match price_sort_direction {
        PriceSortDirection::LowToHigh => "ASC",
        PriceSortDirection::HighToLow => "DESC",
    };

    let visibility = match visibility {
        Some(visibility) => json!(visibility),
        None => json!(null),
    };

    resolve_aql_vector(
        pool,
        r#"
            FOR category IN @categories
              FOR product,e,p IN INBOUND category GRAPH product_categories
                FILTER @search_all == true ? true : (product.is_published IN [true])
                FILTER @visibility == null ? true : (@visibility IN product.visibility)
                SORT product.price.unit_amount @price_sort_direction

                LET t = FIRST(
                  FOR t IN product.translations
                    FILTER t.name != null AND t.locale == @client_locale
                    RETURN t
                )

                RETURN MERGE(
                  product,
                  { unit_label: DOCUMENT(product.unit_label)[@client_locale] },
                  { name: t.name, description: t.description }
                )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "categories" => categories,
            "search_all" => search_all,
            "visibility" => visibility,
            "price_sort_direction" => sort_direction,
        ],
    )
    .await
}

/// Important note: product should be moved into the archive before deleting it!
pub(in crate::commerce) async fn delete_product(
    pool: &ConnectionPool,
    product_key: &str,
    client_locale: &SupportedLocale,
) -> anyhow::Result<Product> {
    // First, fetch product that is being deleted so we can return it later:
    let deleted_product = resolve_aql(
        pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@client_locale]
            LET product = DOCUMENT(products, @product_key)

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @client_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_key" => product_key,
        ],
    )
    .await;

    // Remove the product via Gherial API so the graph edges are not dangling:
    let db = pool.db().await;
    db.remove_graph_vertex(
        "product_categories", // graph
        "products",           // name of the vertex collection the vertex belongs to
        product_key,
        true, // wait for sync
    )
    .await?;

    deleted_product
}
