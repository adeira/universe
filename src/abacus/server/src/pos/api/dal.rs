#[cfg(test)]
use crate::arango::resolve_aql_vector;
use crate::arango::{resolve_aql, ConnectionPool};
use crate::price::SupportedCurrency;
use serde::{Deserialize, Serialize};
use std::fmt::{Debug, Formatter};

#[derive(Serialize, Deserialize, Clone)]
pub struct PosCheckout {
    _id: String,
    created_date: String,
    selected_products: Vec<PosCheckoutProductInput>,
}

impl Debug for PosCheckout {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PosCheckout")
            .field("selected_products", &self.selected_products)
            .finish_non_exhaustive()
    }
}

impl PosCheckout {
    pub fn id(&self) -> String {
        self._id.to_owned()
    }
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub(in crate::pos) struct PosCheckoutProductAddonInput {
    pub(crate) product_addon_id: String,
    pub(crate) product_addon_extra_price_unit_amount: i32,
    pub(crate) product_addon_extra_price_unit_amount_currency: SupportedCurrency,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub(in crate::pos) struct PosCheckoutProductInput {
    // Original product ID just in case it's needed one day (even though it might be already
    // deleted). Already deleted product should not be a big deal since we copy here all the
    // necessary product info at the time of the checkout (so current edited product might be
    // completely different and that's OK).
    pub(crate) product_id: String,
    pub(crate) product_name: String,
    pub(crate) product_units: i32,
    pub(crate) product_price_unit_amount: i32,
    pub(crate) product_price_unit_amount_currency: SupportedCurrency,
    pub(crate) product_addons: Option<Vec<PosCheckoutProductAddonInput>>,
}

#[derive(Serialize, Deserialize)]
pub(in crate::pos) struct PosCheckoutInput {
    pub(crate) selected_products: Vec<PosCheckoutProductInput>,
}

/// Saves checkout information into database so we know what sales are happening via POS and we can
/// make further decisions based on that. Specifically, we record the following information:
///
/// - date and time of the POS sale
/// - products sold (not only IDs but the whole expanded products so future changes of these
///   products don't affect this POS history)
/// - price for each product at the time of the sale (again, preserving historic state)
pub(in crate::pos) async fn create_checkout(
    pool: &ConnectionPool,
    input: &PosCheckoutInput,
) -> anyhow::Result<PosCheckout> {
    resolve_aql(
        pool,
        r#"
            INSERT {
              created_date: DATE_ISO8601(DATE_NOW()),
              selected_products: @selected_products
            } INTO pos_checkouts
            RETURN NEW
        "#,
        hashmap_json![
            "selected_products" => input.selected_products,
        ],
    )
    .await
}

#[cfg(test)]
pub(in crate::pos) async fn get_all_checkouts(
    pool: &ConnectionPool,
) -> anyhow::Result<Vec<PosCheckout>> {
    resolve_aql_vector(
        &pool,
        r#"
            FOR checkout IN pos_checkouts
            RETURN checkout
        "#,
        hashmap_json![],
    )
    .await
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::arango::{cleanup_test_database, prepare_empty_test_database};

    #[ignore]
    #[tokio::test]
    async fn create_checkout_test() {
        let db_name = "create_checkout_test";
        let pool = prepare_empty_test_database(db_name).await;

        // 1) create first checkout
        create_checkout(
            &pool,
            &PosCheckoutInput {
                selected_products: vec![
                    PosCheckoutProductInput {
                        product_id: String::from("products/1"),
                        product_name: String::from("Product name 1"),
                        product_units: 1,
                        product_price_unit_amount: 120,
                        product_price_unit_amount_currency: SupportedCurrency::MXN,
                        product_addons: None,
                    },
                    PosCheckoutProductInput {
                        product_id: String::from("products/2"),
                        product_name: String::from("Product name 2"),
                        product_units: 2,
                        product_price_unit_amount: 150,
                        product_price_unit_amount_currency: SupportedCurrency::MXN,
                        product_addons: None,
                    },
                ],
            },
        )
        .await
        .unwrap();

        // 2) create second checkout
        create_checkout(
            &pool,
            &PosCheckoutInput {
                selected_products: vec![PosCheckoutProductInput {
                    product_id: String::from("products/3"),
                    product_name: String::from("Product name 3"),
                    product_units: 5,
                    product_price_unit_amount: 200,
                    product_price_unit_amount_currency: SupportedCurrency::MXN,
                    product_addons: None,
                }],
            },
        )
        .await
        .unwrap();

        // 3) return the checkout stats
        let all_checkouts = get_all_checkouts(&pool).await.unwrap();
        insta::assert_ron_snapshot!(all_checkouts, {
            "[]._id" => "[REDACTED]",
            "[].created_date" => "[REDACTED]",
        });

        cleanup_test_database(db_name).await;
    }

    #[ignore]
    #[tokio::test]
    async fn create_checkout_with_addons_test() {
        let db_name = "create_checkout_with_addons_test";
        let pool = prepare_empty_test_database(db_name).await;

        // 1) create first checkout
        create_checkout(
            &pool,
            &PosCheckoutInput {
                selected_products: vec![
                    PosCheckoutProductInput {
                        product_id: String::from("products/1"),
                        product_name: String::from("Product name 1"),
                        product_units: 1,
                        product_price_unit_amount: 100,
                        product_price_unit_amount_currency: SupportedCurrency::MXN,
                        product_addons: None,
                    },
                    PosCheckoutProductInput {
                        product_id: String::from("products/1"),
                        product_name: String::from("Product name 1 - with addons"),
                        product_units: 1,
                        product_price_unit_amount: 100,
                        product_price_unit_amount_currency: SupportedCurrency::MXN,
                        product_addons: Some(vec![
                            PosCheckoutProductAddonInput {
                                product_addon_id: String::from("product_addons/1"),
                                product_addon_extra_price_unit_amount: 10,
                                product_addon_extra_price_unit_amount_currency:
                                    SupportedCurrency::MXN,
                            },
                            PosCheckoutProductAddonInput {
                                product_addon_id: String::from("product_addons/2"),
                                product_addon_extra_price_unit_amount: 20,
                                product_addon_extra_price_unit_amount_currency:
                                    SupportedCurrency::MXN,
                            },
                        ]),
                    },
                ],
            },
        )
        .await
        .unwrap();

        // 2) return the checkout stats
        let all_checkouts = get_all_checkouts(&pool).await.unwrap();
        insta::assert_ron_snapshot!(all_checkouts, {
            "[]._id" => "[REDACTED]",
            "[].created_date" => "[REDACTED]",
        });

        cleanup_test_database(db_name).await;
    }
}
