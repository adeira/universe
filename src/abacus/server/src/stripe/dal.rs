use crate::arango::{resolve_aql, ConnectionPool, Document};
use crate::stripe::webhook::StripeWebhookPayload;

pub(in crate) async fn record_webhook_call(
    pool: &ConnectionPool,
    stripe_webhook_payload: &StripeWebhookPayload,
) -> anyhow::Result<Document<StripeWebhookPayload>> {
    resolve_aql(
        pool,
        r#"
            INSERT {
              _key: @webhook_key,
              id: @webhook_key,
              api_version: @webhook_api_version,
              created: @webhook_created,
              type: @webhook_type,
              data: @webhook_data,
            } INTO webhook_events_stripe
            RETURN NEW
        "#,
        hashmap_json![
            "webhook_key" => stripe_webhook_payload.id,
            "webhook_api_version" => stripe_webhook_payload.api_version,
            "webhook_created" => stripe_webhook_payload.created,
            "webhook_type" => stripe_webhook_payload.r#type,
            "webhook_data" => stripe_webhook_payload.data,
        ],
    )
    .await
}
