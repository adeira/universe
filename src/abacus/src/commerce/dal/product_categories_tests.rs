use crate::arangodb::{cleanup_test_database, prepare_empty_test_database};
use crate::commerce::dal::product_categories::{
    get_product_categories_by_ids, search_all_product_categories,
};
use crate::locale::SupportedLocale;

// (cd src/abacus && cargo t -- --ignored commerce::dal::product_categories_tests)

#[ignore]
#[tokio::test]
async fn search_all_product_categories_test() {
    let db_name = "search_all_product_categories_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) try to find all product categories in ENGLISH
    let found_categories_en = search_all_product_categories(&pool, &SupportedLocale::EnUS)
        .await
        .unwrap();

    assert_eq!(found_categories_en.len(), 2);
    insta::assert_debug_snapshot!(found_categories_en);

    // 2) try to find all product categories in SPANISH
    let found_categories_en = search_all_product_categories(&pool, &SupportedLocale::EsMX)
        .await
        .unwrap();

    assert_eq!(found_categories_en.len(), 2);
    insta::assert_debug_snapshot!(found_categories_en);

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn get_product_categories_by_ids_test() {
    let db_name = "get_product_categories_by_ids_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) try to find product categories in ENGLISH by IDs
    let found_categories_en = get_product_categories_by_ids(
        &pool,
        &SupportedLocale::EnUS,
        &[
            "product_categories/coffee".to_string(), // added by DB migration
            "product_categories/tea".to_string(),    // added by DB migration
        ],
    )
    .await
    .unwrap();

    assert_eq!(found_categories_en.len(), 2);
    insta::assert_debug_snapshot!(found_categories_en);

    // 2) try to find one product category in SPANISH using DB key
    let found_categories_en = get_product_categories_by_ids(
        &pool,
        &SupportedLocale::EsMX,
        &[
            "coffee".to_string(), // notice it's not a DB `_id` but DB `_key`
        ],
    )
    .await
    .unwrap();

    assert_eq!(found_categories_en.len(), 1);
    insta::assert_debug_snapshot!(found_categories_en);

    cleanup_test_database(&db_name).await;
}
