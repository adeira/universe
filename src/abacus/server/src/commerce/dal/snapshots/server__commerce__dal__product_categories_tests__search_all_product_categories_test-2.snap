---
source: src/commerce/dal/product_categories_tests.rs
expression: found_categories_en
---
[
    Some(
        ProductCategory {
            name: Some(
                "café",
            ),
            translations: [
                ProductCategoryMultilingualTranslations {
                    locale: EnUS,
                    name: "coffee",
                },
                ProductCategoryMultilingualTranslations {
                    locale: EsMX,
                    name: "café",
                },
            ],
        },
    ),
    Some(
        ProductCategory {
            name: Some(
                "té",
            ),
            translations: [
                ProductCategoryMultilingualTranslations {
                    locale: EnUS,
                    name: "tea",
                },
                ProductCategoryMultilingualTranslations {
                    locale: EsMX,
                    name: "té",
                },
            ],
        },
    ),
    Some(
        ProductCategory {
            name: Some(
                "dumplings",
            ),
            translations: [
                ProductCategoryMultilingualTranslations {
                    locale: EnUS,
                    name: "dumplings",
                },
                ProductCategoryMultilingualTranslations {
                    locale: EsMX,
                    name: "dumplings",
                },
            ],
        },
    ),
    Some(
        ProductCategory {
            name: Some(
                "otro",
            ),
            translations: [
                ProductCategoryMultilingualTranslations {
                    locale: EnUS,
                    name: "other",
                },
                ProductCategoryMultilingualTranslations {
                    locale: EsMX,
                    name: "otro",
                },
            ],
        },
    ),
]
